src/Components/Web/src/WebRenderer.cs

```diff
@@ -40,9 +41,15 @@ public abstract class WebRenderer : Renderer
 
         // Supply a DotNetObjectReference to JS that it can use to call us back for events etc.
         jsComponentInterop.AttachToRenderer(this);
+
         var jsRuntime = serviceProvider.GetRequiredService<IJSRuntime>();
+        var jsRuntimeJsonSerializerOptions = jsRuntime.CloneJsonSerializerOptions();
+        jsRuntimeJsonSerializerOptions.TypeInfoResolverChain.Insert(0, JsonConverterFactoryTypeInfoResolver<DotNetObjectReference<WebRendererInteropMethods>>.Instance);
+        jsRuntimeJsonSerializerOptions.TypeInfoResolverChain.Insert(0, WebRendererSerializerContext.Default);
```

Comment:
It's scoped, not singleton. The scope is per HTTP request for SSR, or per WebSocket connection for Blazor Server.

It might be possible to share a JSO across scopes as long as there's no risk of interference or information leakage across users when doing that.

Reply:
That shouldn't be an issue unless user info is encapsulated by any custom converters.

Reply:
Actually you reminded me that the custom converters very much *do* require access to scoped context. For example see `DotNetObjectReferenceJsonConverter` which has to access the scope's `JSRuntime` in order to keep track of the references on a per-scope basis.

Reply:
Yeah, I think we've had this discussion before and it seems to be an unfortunate consequence of STJ not supporting DI that is scoped to a particular serialization operation ([tracking issue](https://github.com/dotnet/runtime/issues/63795)). Is there any way we could decouple the converters from user state using APIs we have in our disposal today? Perhaps `AsyncLocal<T>`?

Reply:
Given the degree of antipathy the team has towards `AsyncLocal` in general, and the likely back-compat problems with such a change, it's unlikely we'd go that way I'm afraid!